### PR TITLE
Parser: identify lines earlier

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Unify line-finding by doing it at the beginning of parsing. See: https://www.pivotaltracker.com/story/show/183797744

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: lib/rust/parser/debug/tests/parse.rs [hunk 1]
```diff
@@ -48,7 +48,7 @@ macro_rules! test {
 
 #[test]
 fn nothing() {
-    test("", block![]);
+    test("", block![()]);
```
Comment (JaroslavTulach) [inline lib/rust/parser/debug/tests/parse.rs:51]: Does this mean that each EOF is a new line?

File: lib/rust/parser/debug/tests/parse.rs [hunk 2]
```diff
@@ -107,7 +107,7 @@ fn else_block() {
 
 #[test]
 fn plain_comments() {
-    test!("# a b c", ());
+    test!("# a b c", ()());
```
Comment (JaroslavTulach) [inline lib/rust/parser/debug/tests/parse.rs:110]: EOF after after comment yields a new line.

File: lib/rust/parser/debug/tests/parse.rs [hunk 3]
```diff
@@ -107,7 +107,7 @@ fn else_block() {
...
     test!("main = # define main\n 4",
```
Comment (JaroslavTulach) [inline lib/rust/parser/debug/tests/parse.rs:111]: There no additional new line in this case as the new line was already there.

**Review by JaroslavTulach — APPROVED** [review]

File: lib/rust/parser/debug/tests/parse.rs (hunk 1 above)
Comment (kazcw) [reply lib/rust/parser/debug/tests/parse.rs:51]: This is just a change in the answer to "does an empty file contain a single empty line, or no lines". A bit of a philosophical question, so I'm just doing whatever doesn't require writing a special case for it.

File: lib/rust/parser/debug/tests/parse.rs (hunk 2 above)
Comment (kazcw) [reply lib/rust/parser/debug/tests/parse.rs:110]: Actually a change in beginning-of-file behavior: Before I was only creating an implicit `Newline` token to start the first line if the file didn't start with whitespace/ignored tokens. For simplicity, I now just create the token unconditionally.

**Pull request merged**